File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/AddLoadOfCattleActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import android.app.DatePickerDialog;

import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;

import com.google.firebase.database.DatabaseReference;
import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.dataLoaders.cache.holdingLoad.InsertHoldingLoad;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.load.InsertLoadEntity;
import com.trevorwiebe.trackacow.data.db.entities.LoadEntity;
import com.trevorwiebe.trackacow.data.db.holdingUpdateEntities.HoldingLoadEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.Utility;

import java.util.Calendar;

public class AddLoadOfCattleActivity extends AppCompatActivity {

    private TextInputEditText mHeadCount;
    private TextInputEditText mDate;
    private TextInputEditText mMemo;
    private Button mSaveButton;

    private Calendar mCalendar = Calendar.getInstance();
    private DatePickerDialog.OnDateSetListener mDatePicker;
    private String mLotId;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_load_of_cattle);

        this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_white_24dp);
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mLotId = getIntent().getStringExtra("lotId");

        mHeadCount = findViewById(R.id.load_head_count);
        mDate = findViewById(R.id.load_date);
        mMemo = findViewById(R.id.load_memo);
        mSaveButton = findViewById(R.id.save_load_of_cattle);

        String date = Utility.convertMillisToDate(System.currentTimeMillis());
        mDate.setText(date);

        mDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new DatePickerDialog(AddLoadOfCattleActivity.this,
                        mDatePicker,
                        mCalendar.get(Calendar.YEAR),
                        mCalendar.get(Calendar.MONTH),
                        mCalendar.get(Calendar.DAY_OF_MONTH))
                        .show();
            }
        });
        mDatePicker = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
                mCalendar.set(Calendar.YEAR, year);
                mCalendar.set(Calendar.MONTH, month);
                mCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                mDate.setText(Utility.convertMillisToDate(mCalendar.getTimeInMillis()));
            }
        };

        mSaveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (mHeadCount.length() == 0 || mDate.length() == 0) {
                    Snackbar.make(v, "Please fill the blanks", Snackbar.LENGTH_LONG).show();
                } else {
                    DatabaseReference loadPushRef = Constants.BASE_REFERENCE.child(LoadEntity.LOAD).push();

                    int totalHead = Integer.parseInt(mHeadCount.getText().toString());

                    mCalendar.set(Calendar.HOUR_OF_DAY, 0);
                    mCalendar.set(Calendar.MINUTE, 0);
                    mCalendar.set(Calendar.SECOND, 0);
                    mCalendar.set(Calendar.MILLISECOND, 0);

                    long longDate = mCalendar.getTimeInMillis();
                    String loadDescription = mMemo.getText().toString();
                    String loadId = loadPushRef.getKey();

                    LoadEntity loadEntity = new LoadEntity(totalHead, longDate, loadDescription, mLotId, loadId);

                    if (Utility.haveNetworkConnection(AddLoadOfCattleActivity.this)) {
                        loadPushRef.setValue(loadEntity);
                    } else {

                        Utility.setNewDataToUpload(AddLoadOfCattleActivity.this, true);

                        HoldingLoadEntity holdingLoadEntity = new HoldingLoadEntity(loadEntity, Constants.INSERT_UPDATE);
                        new InsertHoldingLoad(holdingLoadEntity).execute(AddLoadOfCattleActivity.this);
                    }

                    new InsertLoadEntity(loadEntity).execute(AddLoadOfCattleActivity.this);

                    setResult(RESULT_OK);
                    finish();
                }
            }
        });

    }

    @Override
    public boolean onSupportNavigateUp() {
        setResult(RESULT_CANCELED);
        finish();
        return false;
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/add_load_of_cattle/AddLoadOfCattleActivity.kt
package com.trevorwiebe.trackacow.presentation.add_load_of_cattle

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.os.Build
import android.os.Bundle
import android.widget.Button
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputEditText
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.models.compound_model.PenAndLotModel
import com.trevorwiebe.trackacow.domain.models.load.LoadModel
import com.trevorwiebe.trackacow.domain.utils.Utility
import dagger.hilt.android.AndroidEntryPoint
import java.util.*

@AndroidEntryPoint
class AddLoadOfCattleActivity : AppCompatActivity() {

    private val mAddLoadOfCattleViewModel: AddLoadOfCattleViewModel by viewModels()

    private lateinit var mHeadCount: TextInputEditText
    private lateinit var mDate: TextInputEditText
    private lateinit var mMemo: TextInputEditText
    private lateinit var mSaveButton: Button

    private val mCalendar = Calendar.getInstance()
    private var mDatePicker: OnDateSetListener? = null
    private var mPenAndLotModel: PenAndLotModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_load_of_cattle)

        this.supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)
        this.supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        @Suppress("DEPRECATION")
        mPenAndLotModel = if (Build.VERSION.SDK_INT >= 33) {
            intent.getParcelableExtra("penAndLotModel", PenAndLotModel::class.java)
        } else {
            intent.getParcelableExtra("penAndLotModel")
        }

        mHeadCount = findViewById(R.id.load_head_count)
        mDate = findViewById(R.id.load_date)
        mMemo = findViewById(R.id.load_memo)
        mSaveButton = findViewById(R.id.save_load_of_cattle)
        val date = Utility.convertMillisToDate(System.currentTimeMillis())
        mDate.setText(date)
        mDate.setOnClickListener {
            DatePickerDialog(
                this@AddLoadOfCattleActivity,
                mDatePicker,
                mCalendar[Calendar.YEAR],
                mCalendar[Calendar.MONTH],
                mCalendar[Calendar.DAY_OF_MONTH]
            )
                .show()
        }
        mDatePicker = OnDateSetListener { view, year, month, dayOfMonth ->
            mCalendar[Calendar.YEAR] = year
            mCalendar[Calendar.MONTH] = month
            mCalendar[Calendar.DAY_OF_MONTH] = dayOfMonth
            mDate.setText(Utility.convertMillisToDate(mCalendar.timeInMillis))
        }
        mSaveButton.setOnClickListener { v ->
            if (mHeadCount.length() == 0 || mDate.length() == 0) {
                Snackbar.make(v, "Please fill the blanks", Snackbar.LENGTH_LONG).show()
            } else {
                val totalHead = mHeadCount.text.toString().toInt()
                mCalendar[Calendar.HOUR_OF_DAY] = 0
                mCalendar[Calendar.MINUTE] = 0
                mCalendar[Calendar.SECOND] = 0
                mCalendar[Calendar.MILLISECOND] = 0
                val longDate = mCalendar.timeInMillis
                val loadDescription = mMemo.text.toString()
                val lotId = mPenAndLotModel?.lotCloudDatabaseId

                val loadModel = LoadModel(0, totalHead, longDate, loadDescription, lotId, "")
                mAddLoadOfCattleViewModel.onEvent(AddLoadOfCattleEvents.OnLoadCreated(loadModel))

                setResult(RESULT_OK)
                finish()
            }
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        setResult(RESULT_CANCELED)
        finish()
        return false
    }
}